crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -1363,13 +1363,27 @@ where
                 }
             }
 
-            ast::Stmt::Assert(node) => {
-                self.visit_expr(&node.test);
-                let predicate = self.record_expression_narrowing_constraint(&node.test);
-                self.record_visibility_constraint(predicate);
+            ast::Stmt::Assert(ast::StmtAssert {
+                test,
+                msg,
+                range: _,
+            }) => {
+                self.visit_expr(test);
 
-                if let Some(msg) = &node.msg {
+                if let Some(msg) = msg {
+                    let pre_assertion = self.flow_snapshot();
```

Comment:
I'm trying to understand why it is correct to take this snapshot after visiting the expression? Should it also be renamed to `post_test` or `pre_msg`?

Reply:
Ah, great point. Yes, this is a snapshot of the state after we've visited the `test` but before we've visited the `msg`. It's the state we need to return to after we've visited the `msg`, since all code following the `assert` statement can only be reached if `test` evaluated to `True` (which means that the `msg` expression would not have been evaluated). I'll rename the variable.